### PR TITLE
Track fetch exceptions for shard follow tasks

### PR DESCRIPTION
**Pull request by jasontedor**
This commit adds tracking and reporting for fetch exceptions. We track fetch exceptions per fetch, keeping track of up to the maximum number of concurrent fetches. With each failing fetch, we associate the from sequence number with the exception that caused the fetch. We report these in the CCR stats endpoint, and add some testing for this tracking.

Relates #30086

### CODE REVIEW

**Comment by jasontedor** [comment]
Here is what a version of the response from this API looks like:

```
{
  "j" : [
    {
      "shard_id" : 0,
      "leader_global_checkpoint" : 1023,
      "leader_max_seq_no" : 1023,
      "follower_global_checkpoint" : 1023,
      "follower_max_seq_no" : 1023,
      "last_requested_seq_no" : 1023,
      "number_of_concurrent_reads" : 1,
      "number_of_concurrent_writes" : 0,
      "number_of_queued_writes" : 0,
      "index_metadata_version" : 7,
      "total_fetch_time_millis" : 17,
      "number_of_successful_fetches" : 1,
      "number_of_failed_fetches" : 9,
      "operations_received" : 0,
      "total_transferred_bytes" : 0,
      "total_index_time_millis" : 0,
      "number_of_successful_bulk_operations" : 0,
      "number_of_failed_bulk_operations" : 0,
      "number_of_operations_indexed" : 0,
      "fetch_errors" : [
        {
          "from_seq_no" : 1024,
          "reason" : {
            "type" : "exception",
            "reason" : "NoShardAvailableActionException[No shard available for [Request={fromSeqNo=1024, maxOperationsCount=1024, shardId=[i][0], maxOperationsSizeInBytes=9223372036854775807}]; nested: RemoteTransportException[[p_hIWo8][127.0.0.1:9300][indices:data/read/xpack/ccr/shard_changes[s]]]; nested: IndexNotFoundException[no such index];",
            "caused_by" : {
              "type" : "no_shard_available_action_exception",
              "reason" : "No shard available for [Request={fromSeqNo=1024, maxOperationsCount=1024, shardId=[i][0], maxOperationsSizeInBytes=9223372036854775807}]",
              "caused_by" : {
                "type" : "index_not_found_exception",
                "reason" : "no such index",
                "index_uuid" : "TfDzmXK_RmaQvyzop98mDA",
                "index" : "i"
              }
            }
          }
        }
      ]
    }
  ]
}
```

Note that I have added a `Object#toString` implementation in this PR for `ShardChangesAction$Request` so that we have a more useful output in the above when formatting the cause.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java
Comment (martijnvg) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java]: this line is too long, causing a check style violation

File: test/framework/src/main/java/org/elasticsearch/test/AbstractSerializingTestCase.java
Comment (martijnvg) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractSerializingTestCase.java]: Ah this was just missing in this base class. It does exist in the `AbstractXContentTestCase` base class.
Should we add this change back into master / 6.x separately after this PR is merged or just wait for when the ccr branches are merged? I have no strong opinion here.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java
Comment (martijnvg) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java]: 👍 to the approach to keep track of exceptions by from seqno in a fixed size linked hashmap.

**Review by martijnvg — APPROVED** [review]
LGTM

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java
Comment (jasontedor) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java]: Thanks. I pushed 477c3bf5e7bb8fd823bfd566bb95418609058c00.

File: test/framework/src/main/java/org/elasticsearch/test/AbstractSerializingTestCase.java
Comment (jasontedor) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractSerializingTestCase.java]: We discussed this offline. We will keep this in this branch for now, and it will come to 6.x/master when we integrate there.

**Comment by jasontedor** [comment]
This PR needs #33113 first.

**Comment by jasontedor** [comment]
@elasticmachine test this please

**Pull request merged**
